drivers/flash/soc_flash_nrf_mramc.c

```diff
@@ -0,0 +1,161 @@
+#define _EBS(node_id)	   DT_PROP(node_id, erase_block_size),
+
+#define MRAM_SIZE   (0 DT_INST_FOREACH_CHILD_STATUS_OKAY(0, _ADD_SIZE))
+#define ERASE_VALUE (uint8_t)NRFY_MRAMC_WORD_AFTER_ERASED
+
+/* write block size and erase block size is in the child node of
+ * mramc DT_FOREACH_CHILD_STATUS_OKAY is used to get the values
+ * from the first child node.
+ */
+static const uint32_t write_block_size_arr[] = {
+	DT_INST_FOREACH_CHILD_STATUS_OKAY(0, _WBS)
+};
+
+static const uint32_t erase_block_size_arr[] = {
+	DT_INST_FOREACH_CHILD_STATUS_OKAY(0, _EBS)
+};
+
+static const uint32_t write_block_size = write_block_size_arr[0];
+static const uint32_t erase_block_size = erase_block_size_arr[0];
+
+static int nrf_mramc_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+	uint32_t addr = NRFX_MRAMC_MAP_TO_ADDR(offset);
+
+	LOG_DBG("read: %x:%zu", addr, len);
+	nrfx_mramc_buffer_read(data, addr, len);
+	return 0;
+}
+
+static int nrf_mramc_write(const struct device *dev, off_t offset,
+			 const void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+	uint32_t addr = NRFX_MRAMC_MAP_TO_ADDR(offset);
+
+	if (len % write_block_size != 0) {
+		LOG_ERR("Write len must be multiple of write_block_size:%zu",
+			 len);
+		return -EINVAL;
+	}
+
+	LOG_DBG("write: %x:%zu", addr, len);
+	nrfx_mramc_words_write(addr, data, len / sizeof(uint32_t));
+
+	return 0;
+}
+
+static int nrf_mramc_erase(const struct device *dev, off_t offset, size_t size)
+{
+	ARG_UNUSED(dev);
+	uint32_t addr = NRFX_MRAMC_MAP_TO_ADDR(offset);
+
+	LOG_DBG("erase: %p:%zu", (void *)addr, size);
+
+	nrfx_mramc_area_erase(addr, size / write_block_size);
+#if defined(NRF_ICACHE)
+	if (nrf_cache_enable_check(NRF_ICACHE)) {
+		while (nrf_cache_busy_check(NRF_ICACHE)) {
+		}
+		nrf_cache_invalidate(NRF_ICACHE);
+	}
+#endif
+	return 0;
+}
+
+static int nrf_mramc_get_size(const struct device *dev, uint64_t *size)
+{
+	ARG_UNUSED(dev);
+	*size = nrfx_mramc_memory_size_get();
+	return 0;
+}
+
+static const struct flash_parameters *nrf_mramc_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	static const struct flash_parameters parameters = {
+		.write_block_size = write_block_size,
+		.erase_value = ERASE_VALUE,
+		.caps = {
+			.no_explicit_erase = true,
+		},
+	};
+
+	return &parameters;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void nrf_mramc_page_layout(const struct device *dev,
+				 const struct flash_pages_layout **layout, size_t *layout_size)
+{
+	ARG_UNUSED(dev);
+
+	static const struct flash_pages_layout pages_layout = {
+		.pages_count = MRAM_SIZE / (erase_block_size),
+		.pages_size = erase_block_size,
+	};
+
+	*layout = &pages_layout;
+	*layout_size = 1;
+}
+#endif
+
+static int mramc_sys_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	if ((erase_block_size % write_block_size) != 0) {
+		LOG_ERR("Erase block size (%d) must be a multiple of write block size (%d)",
+				erase_block_size, write_block_size);
+		return -EINVAL;
+	}
```

Comment:
Can this be figured out at compile time? This is taken from DTS, so I guess it can.